paymentsheet/src/main/java/com/stripe/android/paymentsheet/viewmodels/BaseSheetViewModel.kt

```diff
@@ -196,15 +210,47 @@ internal abstract class BaseSheetViewModel<TransitionTargetType>(
         }
     }
 
-    fun getSupportedPaymentMethods(): List<SupportedPaymentMethod> {
-        stripeIntent.value?.let { stripeIntent ->
-            return stripeIntent.paymentMethodTypes.mapNotNull {
+    private fun getSupportedPaymentMethods(
+        stripeIntentParameter: StripeIntent?
+    ): List<SupportedPaymentMethod> {
+        stripeIntentParameter?.let { stripeIntent ->
+            return stripeIntent.paymentMethodTypes.asSequence().mapNotNull {
                 SupportedPaymentMethod.fromCode(it)
             }.filterNot {
                 // AfterpayClearpay requires a shipping address, filter it out if not provided
-                it == SupportedPaymentMethod.AfterpayClearpay &&
+                val excludeAfterPay = it == SupportedPaymentMethod.AfterpayClearpay &&
                     (stripeIntent as? PaymentIntent)?.shipping == null
+                if (excludeAfterPay) {
+                    logger.debug(
+                        "AfterPay will not be shown. It requires that Shipping is " +
+                            "included in the Payment or Setup Intent"
+                    )
+                }
+                excludeAfterPay
+            }.filterNot { supportedPaymentMethod ->
+                val excludeRequiresMandate =
+                    (stripeIntent is SetupIntent) && supportedPaymentMethod.requiresMandate
+                if (excludeRequiresMandate) {
+                    logger.debug(
+                        "${supportedPaymentMethod.name} will not be shown. It " +
+                            "requires a mandate which is incompatible with SetupIntents"
+                    )
+                }
+                excludeRequiresMandate
+            }.filterNot { supportedPaymentMethod ->
+                val excludeRequiresMandate = (stripeIntent is PaymentIntent) &&
+                    supportedPaymentMethod.requiresMandate &&
+                    stripeIntent.setupFutureUsage == StripeIntent.Usage.OffSession
+                if (excludeRequiresMandate) {
+                    logger.debug(
+                        "${supportedPaymentMethod.name} will not be shown.  It " +
+                            "requires a mandate which is incompatible with off_session " +
+                            "PaymentIntents"
+                    )
+                }
+                excludeRequiresMandate
```

Comment:
We'll need to filter these from the saved PMs too, right?

Reply:
This happens here: https://github.com/stripe/stripe-android/pull/4191/files#r713334571